conjure-go-client/httpclient/authn_test.go

```diff
@@ -113,5 +116,205 @@ func TestRoundTripperWithBasicAuthProvider(t *testing.T) {
+	apiTokenFile := filepath.Join(t.TempDir(), "token.txt")
+	require.NoError(t, os.WriteFile(apiTokenFile, []byte(token), 0600))
+
+	noAuthServer := func(t *testing.T) *httptest.Server {
+		return httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			if assert.NotContains(t, req.Header, "Authorization") {
+				rw.WriteHeader(http.StatusOK)
+			} else {
+				rw.WriteHeader(http.StatusUnauthorized)
+			}
+		}))
+	}
+	basicAuthServer := func(t *testing.T) *httptest.Server {
+		basicAuthHeader := "Basic " + base64.StdEncoding.EncodeToString([]byte(username+":"+password))
+		return httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			if assert.Equal(t, basicAuthHeader, req.Header.Get("Authorization")) {
+				rw.WriteHeader(http.StatusOK)
+			} else {
+				rw.WriteHeader(http.StatusUnauthorized)
+			}
+		}))
+	}
+	bearerAuthServer := func(t *testing.T) *httptest.Server {
+		bearerAuthHeader := "Bearer " + token
+		return httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			if assert.Equal(t, bearerAuthHeader, req.Header.Get("Authorization")) {
+				rw.WriteHeader(http.StatusOK)
+			} else {
+				rw.WriteHeader(http.StatusUnauthorized)
+			}
+		}))
+	}
+
+	for _, tc := range []struct {
+		Name          string
+		Server        func(t *testing.T) *httptest.Server
+		Config        httpclient.ClientConfig
+		ClientParams  []httpclient.ClientOrHTTPClientParam
+		RequestParams []httpclient.RequestParam
+	}{
+		{
+			Name:   "NoAuth",
+			Server: noAuthServer,
+		},
+		// Basic Auth
+		{
+			Name:   "BasicAuth config",
+			Server: basicAuthServer,
+			Config: httpclient.ClientConfig{BasicAuth: &httpclient.BasicAuth{User: username, Password: password}},
+		},
+		{
+			Name:         "WithBasicAuth param",
+			Server:       basicAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuth(username, password)},
+		},
+		{
+			Name:   "WithBasicAuthProvider param",
+			Server: basicAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuthProvider(func(ctx context.Context) (httpclient.BasicAuth, error) {
+				return httpclient.BasicAuth{User: username, Password: password}, nil
+			})},
+		},
+		{
+			Name:   "WithBasicAuthOptionalProvider present",
+			Server: basicAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuthOptionalProvider(func(ctx context.Context) (*httpclient.BasicAuth, error) {
+				return &httpclient.BasicAuth{User: username, Password: password}, nil
+			})},
+		},
+		{
+			Name:   "WithBasicAuthOptionalProvider absent",
+			Server: noAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuthOptionalProvider(func(ctx context.Context) (*httpclient.BasicAuth, error) {
+				return nil, nil
+			})},
+		},
+		{
+			Name:         "WithBasicAuth param beats basic config",
```

Comment:
this test fails on develop because the header set by WithBasicAuth is overwritten by the client's middleware